SO-2929: Report error on relationships with inactive source, destination or type concepts

zoliMolnar opened the pull request:
This pull request adds a validation rule to test if a relationship has inactive source, destination or type concept.

https://snowowl.atlassian.net/browse/SO-2929

Commit:
SO-2929: Added utility classes to snomed validation package.

Commit:
SO-2929: Added com.b2international.snowowl.snomed.validation package...

under snowowl repository.

Commit:
SO-2929: Added the validation rules file and the rule groovy script

Commit:
SO-2929: Added test suite class and a base class to index rules and ...

... validate and the Test for the groovy script.

Commit:
SO-2929: Rename rule from rule1 to ruleSnomedCommon1.

Commit:
SO-2929: Rename ruleId from 1 to snomed-common-1

Commit:
SO-2929: Rename validation rules json file.

Commit:
SO-2929: Added support for multiple validation-rules json files.

Commit:
SO-2929: Rename variable.

Commit:
SO-2929: Changed the URL of the json file.

Commit:
SO-2944: Added FSN uniqueness rule as a common rule.

cmark commented on com.b2international.snowowl.validation.snomed/src/main/resources/scripts/ruleSnomedCommon1.groovy:
```diff
@@ -0,0 +1,66 @@
+
+Set<ComponentIdentifier> invalidIds = Sets.newHashSet()
+
+SnomedConcepts inactiveConcepts = SnomedRequests.prepareSearchConcept()
```
Just to save a bunch of memory, return only the IDs of the matching components. In order to that efficiently, use the Index API directly rather than the high-level request API.

cmark commented on com.b2international.snowowl.validation.snomed/src/main/resources/scripts/ruleSnomedCommon2.groovy:
```diff
@@ -0,0 +1,49 @@
+
+Aggregation<SnomedDescriptionIndexEntry> buckets = ctx.service(RevisionSearcher.class)
+		.aggregate(AggregationBuilder
+		.bucket("rule52", SnomedDescriptionIndexEntry.class)
```
rule52 is an NHS specific rule ID, since we've moved this rule to the generic set please update the name of the bucket as well.

cmark commented on com.b2international.snowowl.validation.snomed/src/tests/java/com/b2international/snowowl/validation/snomed/AllGenericValidationTests.java:
```diff
@@ -0,0 +1,34 @@
+/*******************************************************************************
+ * Copyright (c) 2018 B2i Healthcare. All rights reserved.
```
Use the OSS license header instead of the proprietary when pushing changes to this repository.

cmark commented on com.b2international.snowowl.validation.snomed/src/tests/java/com/b2international/snowowl/validation/snomed/BaseGenericValidationRuleTest.java:
```diff
@@ -0,0 +1,156 @@
+	private static final Long EPRESCRIBING_ROUTE_SIMPLE_REFSETL = Long.parseLong("999000051000001100");
+	private static final Long CONCEPT_MODEL_ATTRIBUTEL = Long.parseLong(Concepts.CONCEPT_MODEL_ATTRIBUTE);
+	private static final Long MODULE_ROOTL = Long.parseLong(Concepts.MODULE_ROOT);
+	private static final Long REFSET_ROOTL = Long.parseLong(Concepts.REFSET_ROOT_CONCEPT);
```
Remove all unused IDs from these constants.

cmark commented on com.b2international.snowowl.validation.snomed/src/tests/java/com/b2international/snowowl/validation/snomed/util/DocumentBuilders.java:
```diff
@@ -0,0 +1,108 @@
+ * 
+ * @since 6.4
+ */
+public abstract class DocumentBuilders {
```
If possible and the dependencies allow it, please move this class and all its companions to `test.commons` bundle.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationBootstrap.java:
```diff
@@ -76,48 +80,64 @@ public void preRun(SnowOwlConfiguration configuration, Environment env) throws E
+			
+			final List<ValidationRule> availableRules = Lists.newArrayList();
+			for (File validationRulesFile : validationRuleFiles) {
+				if (validationRulesFile.exists()) {
```
Since you're collecting the files via `File.listFiles()` this check is unnecessary.

Commit:
SO-2929: Moved DocumentBuilders RandomSnomedIdentiferGenerator and...

... TestBranchContext classes into
com.b2international.snowowl.test.commons.snomed package

Commit:
SO-2929: Exported com.b2international.snowowl.test.commons.snomed...

... added com.b2international.snowowl.test.commons.snomed to manifest as
a required bundle.

Commit:
SO-2929: Renamed rule in aggregation

Commit:
SO-2929: Changed license headers to OSS. Removed unnecessary check from

ValidationBootstrap

Commit:
SO-2929: Refactored rule to use ID-s only, and to use scrolling as well.

Commit:
SO-2929: Morphed the expressions in to one.

cmark commented on com.b2international.snowowl.validation.snomed/src/main/resources/scripts/ruleSnomedCommon1.groovy:
```diff
@@ -1,66 +1,53 @@
-}
-
+Iterable<Hits<String>> inactiveConceptHits = ctx.service(RevisionSearcher.class)
+		.scroll(Query.select(String.class)
```
Convert this scroll to a single search with limit `Integer.MAX_VALUE`.

Commit:
SO-2929: Removed scrolling from inactive concept search query.

cmark commented:
LGTM!
